File: web/admin/src/lib/permissions.ts
/**
 * Матрица прав UI — синхронизирована с internal/auth/permissions.go.
 * При изменении backend-матрицы обновить этот файл (Phase 4: codegen).
 */

export const GROUP_STAFF = 'staff';
export const GROUP_ADMIN = 'admin';

export const ROLE_ADMIN = 'admin';
export const ROLE_MANAGER = 'manager';

export const STAFF_ROLES = [ROLE_ADMIN, ROLE_MANAGER] as const;
export const ADMIN_ROLES = [ROLE_ADMIN] as const;

export type StaffRole = (typeof STAFF_ROLES)[number];
export type AdminRole = (typeof ADMIN_ROLES)[number];

/** UI-маршруты и требуемая группа (см. docs/plans/auth-refactor.md). */
const UI_ROUTE_GROUPS: { prefix: string; group: string }[] = [
	{ prefix: '/settings', group: GROUP_ADMIN }
];

/** Публичные маршруты без авторизации. */
export function isPublicPath(pathname: string): boolean {
	return pathname === '/login' || pathname === '/logout' || pathname.startsWith('/logout/');
}

export function getRequiredGroup(pathname: string): string | null {
	if (isPublicPath(pathname)) return null;
	for (const { prefix, group } of UI_ROUTE_GROUPS) {
		if (pathname === prefix || pathname.startsWith(prefix + '/')) return group;
	}
	return GROUP_STAFF;
}

export function hasRole(role: string, allowed: readonly string[]): boolean {
	return allowed.includes(role);
}

export function canAccessRoute(role: string, pathname: string): boolean {
	const group = getRequiredGroup(pathname);
	if (group === null) return true;
	if (group === GROUP_ADMIN) return hasRole(role, ADMIN_ROLES);
	if (group === GROUP_STAFF) return hasRole(role, STAFF_ROLES);
	return false;
}

export function isAdminRole(role: string | undefined | null): boolean {
	return role === ROLE_ADMIN;
}

File: internal/auth/permissions.go
// Package auth defines the role/permission matrix shared by routes and admin UI.
package auth

import "github.com/piplos/piplos.media/internal/models"

// Permission groups (single source of truth for route access).
const (
	GroupPublic        = "public"
	GroupAuthenticated = "authenticated"
	GroupStaff         = "staff"
	GroupAdmin         = "admin"
)

// StaffRoles may access content, uploads, files, leads, read languages, translate.
var StaffRoles = []models.UserRole{models.RoleAdmin, models.RoleManager}

// AdminRoles may access users, settings, language CRUD, AI models, backups.
var AdminRoles = []models.UserRole{models.RoleAdmin}

// AuthenticatedRoles may call /auth/me and /auth/logout.
var AuthenticatedRoles = StaffRoles

// RoutePermission documents an API endpoint and its required group (for codegen/CI).
type RoutePermission struct {
	Method string
	Path   string
	Group  string
}

// APIRoutes is the canonical permission matrix for /v1 endpoints.
var APIRoutes = []RoutePermission{
	// Public
	{Method: "POST", Path: "/v1/leads", Group: GroupPublic},
	{Method: "GET", Path: "/v1/public/*", Group: GroupPublic},
	{Method: "POST", Path: "/v1/auth/login", Group: GroupPublic},
	{Method: "POST", Path: "/v1/auth/refresh", Group: GroupPublic},
	// Authenticated
	{Method: "GET", Path: "/v1/auth/me", Group: GroupAuthenticated},
	{Method: "POST", Path: "/v1/auth/logout", Group: GroupAuthenticated},
	{Method: "GET", Path: "/v1/auth/permissions", Group: GroupAuthenticated},
	// Staff (content, uploads, files, leads, languages read, translate)
	{Method: "*", Path: "/v1/projects", Group: GroupStaff},
	{Method: "*", Path: "/v1/services", Group: GroupStaff},
	{Method: "*", Path: "/v1/stack", Group: GroupStaff},
	{Method: "*", Path: "/v1/seo", Group: GroupStaff},
	{Method: "*", Path: "/v1/pages", Group: GroupStaff},
	{Method: "*", Path: "/v1/legal", Group: GroupStaff},
	{Method: "*", Path: "/v1/uploads", Group: GroupStaff},
	{Method: "*", Path: "/v1/files", Group: GroupStaff},
	{Method: "*", Path: "/v1/leads", Group: GroupStaff},
	{Method: "GET", Path: "/v1/languages", Group: GroupStaff},
	{Method: "POST", Path: "/v1/translate", Group: GroupStaff},
	// Admin only
	{Method: "*", Path: "/v1/users", Group: GroupAdmin},
	{Method: "*", Path: "/v1/settings", Group: GroupAdmin},
	{Method: "POST", Path: "/v1/languages", Group: GroupAdmin},
	{Method: "DELETE", Path: "/v1/languages/*", Group: GroupAdmin},
	{Method: "*", Path: "/v1/ai-models", Group: GroupAdmin},
	{Method: "*", Path: "/v1/backups", Group: GroupAdmin},
}
